python/imzml_reader.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import pandas as pd
+from pyimzml.ImzMLParser import ImzMLParser
+import zarr
+
+import argparse
+
+
+class IMSDataset:
+    """Converts IMS data stored as imzML into columnar or ndarray formats
+
+    :param imzml_file: path to `.imzML` file.
+    :param ibd_file: path to associated `.ibd` file.
+    :param micro_res: microscopy resolution in nm (used for scaling).
+    :param ims_res: IMS resolution in nm (used for scaling).
+    """
+
+    def __init__(self, imzml_file, ibd_file, micro_res, ims_res):
+        # When passing the ibd path explicitly,
+        # the file object must be opened manually
+        self.parser = ImzMLParser(
+            filename=imzml_file, ibd_file=open(ibd_file, "rb")
+        )
+        self.micro_res = micro_res
+        self.ims_res = ims_res
+        self.ims_px_in_micro = ims_res / micro_res  # scaling factor
+
+        mz_lengths = self.parser.mzLengths
+        if not (mz_lengths.count(mz_lengths[0]) == len(mz_lengths)):
+            raise ValueError(
+                "The number of m/z is not the same at each coordinate."
+            )
+        self.mzs, _ = self.parser.getspectrum(0)
+
+    def __get_min_max_coords(self):
+        coords = np.array(self.parser.coordinates)
+        x_min, y_min, _ = np.min(coords, axis=0)
+        x_max, y_max, _ = np.max(coords, axis=0)
+        return x_min, y_min, x_max, y_max
+
+    def __format_mzs(self, precision=4):
+        return np.round(self.mzs, precision).astype(str)
+
+    def to_columnar(self, dtype="uint32"):
+        coords = np.array(self.parser.coordinates)
+        x, y, _ = coords.T
+
+        coords_df = pd.DataFrame(
+            {
+                "x": x,
+                "y": y,
+                "micro_x_topleft": x * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_y_topleft": y * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_px_width": np.repeat(self.ims_px_in_micro, len(coords)),
+            },
+            dtype=dtype,
+        )
+
+        intensities = np.zeros((len(coords_df), len(self.mzs)))
+        for i in range(len(coords)):
+            _, coord_intensities = self.parser.getspectrum(i)
+            intensities[i, :] = coord_intensities
+
+        intensities = pd.DataFrame(
+            intensities, columns=self.__format_mzs(), dtype=dtype,
+        )
+
+        return coords_df.join(intensities)
+
+    def to_array(self):
+        x_min, y_min, x_max, y_max = self.__get_min_max_coords()
+        arr = np.zeros(
+            (self.parser.mzLengths[0], x_max - x_min + 1, y_max - y_min + 1)
+        )
+
+        for idx, (x, y, _) in enumerate(self.parser.coordinates):
+            _, intensities = self.parser.getspectrum(idx)
+            arr[:, x - x_min, y - y_min] = intensities
```

Comment:
Maybe a comprehension could be used here?

Reply:
never mind.

Reply:
Again, I decided to pre-allocate the array before hand as I thought it would be a faster, and here (more so than in `to_columnar`), I'd really need to do some mental gymnastics as to how to do the list comprehension. 

I create the 3D array and then use the shifted x and y coordinates to fill the mz dimension. I also don't know if we can rely on the ordering of the coordinates returned from `self.parser.coordinates`. 

for example:
```python
for idx, coord in enumerate(self.parser.coordinates):
     print(coord)
# (0, 0)      (0, 0)       (1,2)             
# (0, 1)      (1, 0)       (1,0)
# (0, 2)      (0, 1)       (0,2)
# (1, 0)  vs  (1, 1)   vs  (0,0)
# (1, 1)      (0, 2)       (0,1)
# (1, 2)      (1, 2)       (1,2)
```

Reply:
It's difficult because we don't actually store the x and y coordinates in the array but moreover use the coordinates to index and fill the corresponding mz ratios. With the thought I've given it, this seems to me the most reliable way to get the "correct" ordering in the 3D array.

Reply:
Sorry, didn't see your most recent comment until now. Should I add a similar comment here?

Reply:
Sure. (And thank you for taking the time to explain: Sorry to put you to the trouble. You could probably copy and paste part of your reply into a comment in the code, if you felt like.)